### DISCUSSION TITLE
How to include prebuilt wallet core library in Java application?

### DISCUSSION THREAD

**Discussion by roxrook** [trustwallet/wallet-core#2634] [Q&A]
Hi there,

I'd love to integrate `wallet-core` into my Java backend application using the docker approach. Here is what I tried so far:
- Build wallet core locally which generates these files under `build` directory
```
CMakeCache.txt        bin                   lib                   local                 walletconsole
CMakeFiles            cmake_install.cmake   libTrustWalletCore.a  tests
Makefile              compile_commands.json libprotobuf.a         trezor-crypto
```
- Next I copied the `jni` folder into my project and link it using similar technique on Android:
```kt
init {
    System.loadLibrary("TrustWalletCore")
  }
```
- Unfortunately, I ran into error saying the library is not available in java path which make senses because I haven't done anything with the `libTrustWalletCore.a`. 
```
no TrustWalletCore in java.library.path: /Users/roxrook/Library/Java/Extensions:/Library/Java/Extensions:/Network/Library/Java/Extensions:/System/Library/Java/Extensions:/usr/lib/java:.
```
I'm wondering if I'm on the right track here or I have to build the whole project using something similar to `ndk` like Android? The reason I used this approach because first `ndk` is not available on the backend and integrating cmake is also a pain, second I also want to use this for my Go backend and docker seems to be the best approach.

**Comment by miloserdow** [comment]
Hi @roxrook, we are publishing [Maven packages](https://github.com/trustwallet/wallet-core/packages/700258?version=3.0.4) for wallet-core with each release. The simplest solution would be to just use maven.

**Reply by hewigovens** [reply]
To clarify: current binaries on Maven is built by Android NDK, you might need to change CMakeLists.txt to build dynamic library (so) instead

**Reply by roxrook** [reply]
@miloserdow Thanks, I tried maven library but it's `.aar` extension and it's not compatible with Java backend application. 
@hewigovens I'm not sure which file to look, can you give me some pointers? I'm currently look at:
```
# Source files
if (${ANDROID})
    message("Configuring for JNI")
    file(GLOB_RECURSE sources src/*.c src/*.cc src/*.cpp src/*.h jni/cpp/*.c jni/cpp/*.cpp jni/cpp/*.h jni/cpp/*.c)
    add_library(TrustWalletCore SHARED ${sources} ${PROTO_SRCS} ${PROTO_HDRS})
    find_library(log-lib log)
    target_link_libraries(TrustWalletCore PUBLIC ${PROJECT_NAME}_INTERFACE PRIVATE TrezorCrypto protobuf ${log-lib} Boost::boost)
else ()
    message("Configuring standalone")
    file(GLOB_RECURSE sources src/*.c src/*.cc src/*.cpp src/*.h)
    add_library(TrustWalletCore ${sources} ${PROTO_SRCS} ${PROTO_HDRS})

    target_link_libraries(TrustWalletCore PUBLIC ${PROJECT_NAME}_INTERFACE PRIVATE TrezorCrypto protobuf Boost::boost)
endif ()
```
Also the `maven-push.gradle`
```groovy
publications {
    Production(MavenPublication) {
      artifact("$buildDir/outputs/aar/trustwalletcore-release.aar")
      artifact sourcesJar {
        classifier "sources"
      }
      groupId this.group
      artifactId 'wallet-core'
      version this.version

      pom.withXml {
        def dependenciesNode = asNode().appendNode('dependencies')
        configurations.implementation.allDependencies.each {
          if (it.name != 'unspecified') {
            def dependencyNode = dependenciesNode.appendNode('dependency')
            dependencyNode.appendNode('groupId', it.group)
            dependencyNode.appendNode('artifactId', it.name)
            dependencyNode.appendNode('version', it.version)
          }
        }
      }
    }
  }
```
Let me know if I'm on the right track, thanks!

**Reply by hewigovens** [reply]
@Milerius can you advise?

**Reply by Milerius** [reply]
@roxrook Hello, we do not have official sample for desktop/java.

As I can see the CMake already built  TrustWalletCore as a shared library for android.

I can advise the following for you, set up a variable `MY_UNIQUE_VARIABLE_FORCING_JNI` in the CMake in order to enter to the conditional code:

```
# Source files
if (${ANDROID} OR ${MY_UNIQUE_VARIABLE_FORCING_JNI})
    message("Configuring for JNI")
    file(GLOB_RECURSE sources src/*.c src/*.cc src/*.cpp src/*.h jni/cpp/*.c jni/cpp/*.cpp jni/cpp/*.h jni/cpp/*.c)
    add_library(TrustWalletCore SHARED ${sources} ${PROTO_SRCS} ${PROTO_HDRS})
    find_library(log-lib log)
    target_link_libraries(TrustWalletCore PUBLIC ${PROJECT_NAME}_INTERFACE PRIVATE TrezorCrypto protobuf ${log-lib} Boost::boost)
else ()
    message("Configuring standalone")
    file(GLOB_RECURSE sources src/*.c src/*.cc src/*.cpp src/*.h)
    add_library(TrustWalletCore ${sources} ${PROTO_SRCS} ${PROTO_HDRS})

    target_link_libraries(TrustWalletCore PUBLIC ${PROJECT_NAME}_INTERFACE PRIVATE TrezorCrypto protobuf Boost::boost)
endif ()
```

Alternatively, you can remove the `if/else` and just keep the jni part.

**Comment by roxrook** [comment]
@Milerius Thanks. I've tried various approaches but run into various build issues:
- I tried what you suggested above but run into protobuf build issue on AWS linux.
- I also tried to install `Android` on my docker image but run into Java JDK issue, our backend is targeting Java 17+.
I'm not sure what's the best path forward. Given that you guys already have the build release process in place, is it possible to publish another Java library bundled with native lib, similar to `h3` library here? https://github.com/uber/h3-java/blob/master/src/main/c/h3-java/CMakeLists.txt

**Reply by Milerius** [reply]
> @Milerius Thanks. I've tried various approaches but run into various build issues:
> 
> * I tried what you suggested above but run into protobuf build issue on AWS linux.
> * I also tried to install `Android` on my docker image but run into Java JDK issue, our backend is targeting Java 17+.
>   I'm not sure what's the best path forward. Given that you guys already have the build release process in place, is it possible to publish another Java library bundled with native lib, similar to `h3` library here? https://github.com/uber/h3-java/blob/master/src/main/c/h3-java/CMakeLists.txt

Hello @roxrook - could you please post the protobuf build issue on AWS linux you encounter and the CMake solutions that you tried?
